### PR TITLE
API keys

### PR DESCRIPTION
**Pull request by pfeuffer**
## Proposed changes

This adds so called "API keys" aka Tokens a user can create as an alternative for his usual credentials to access the REST API and for scm client access (git, hg and svn).

The essential part of an API key is a random 20 digit passphrase. Together with the user name and an id this is packed as a json representation and encoded as a base64 string. This is presented once to the user (or as the result of the REST api call to create the key), but will be stored nowhere. Only a hashed value of this random passphrase is stored (like the normal user passwords), so that such a token can more or less be checked the same way (after extraction) a password is checked.

For this, a new realm has been added, that takes these api keys either as a bearer token or as a password for basic authentication.

API keys are limited to repositories and a specific role. Therefore they cannot be used for global configuration or user and group administration.

### Your checklist for this pull request

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] PR is well described
- [x] Related issues linked to PR if existing and labels set
- [x] Target branch is not master (in most cases develop should bet the target of choice) 
- [x] Code does not conflict with target branch
- [x] New code is covered with unit tests
- [x] CHANGELOG.md updated
- [x] Definition of Done's fulfilled: [DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/definition-of-done.md) // [UI DoD](https://github.com/scm-manager/scm-manager/blob/develop/docs/en/development/ui-dod.md)
- [x] Documentation updated (only necessary for new features or changed behaviour)

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

File: scm-webapp/src/main/java/sonia/scm/security/ApiKeyWithPassphrase.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import sonia.scm.xml.XmlInstantAdapter;
+
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
+import java.time.Instant;
+
+@AllArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/security/ApiKeyWithPassphrase.java:40]: I just learned something new :+1:

File: scm-webapp/src/main/java/sonia/scm/security/ScmWildcardPermission.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static java.util.Collections.singleton;
+import static java.util.Optional.empty;
+import static java.util.Optional.of;
+
+public class ScmWildcardPermission extends WildcardPermission {
```
Comment (eheimbuch) [inline scm-webapp/src/main/java/sonia/scm/security/ScmWildcardPermission.java:40]: This class is kinda hard to understand. Maybe leave some comments?

File: scm-webapp/src/test/java/sonia/scm/security/ApiKeyTokenHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+class ApiKeyTokenHandlerTest {
+
+  ApiKeyTokenHandler handler = new ApiKeyTokenHandler();
+
+  @Test
+  void shouldSerializeAndDeserializeToken() {
+    final String tokenString = handler.createToken("dent", new ApiKey("42", "hg2g", "READ", now()), "some secret");
+
+    System.out.println(tokenString);
```
Comment (eheimbuch) [inline scm-webapp/src/test/java/sonia/scm/security/ApiKeyTokenHandlerTest.java:43]: We still need this?

File: scm-webapp/src/test/java/sonia/scm/security/ScopesTest.java [hunk 4]
```diff
@@ -21,117 +21,126 @@
...
     Scope scope = Scope.valueOf("repository:read:123");
     AuthorizationInfo authz = authz("repository:*", "user:*:me");
-    
+
     assertPermissions(Scopes.filter(resolver, authz, scope), "repository:read:123");
   }
-  
+
+  @Test
+  public void testFilterX() {
```
Comment (eheimbuch) [inline scm-webapp/src/test/java/sonia/scm/security/ScopesTest.java:74]: What is FilterX?

**Review by eheimbuch — CHANGES_REQUESTED** [review]

File: scm-ui/ui-webapp/src/users/components/apiKeys/AddApiKey.tsx [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  const [error, setError] = useState<undefined | Error>();
+  const [displayName, setDisplayName] = useState("");
+  const [permissionRole, setPermissionRole] = useState("");
+  const [addedKey, setAddedKey] = useState("");
+
+  useEffect(() => {
+    if (!availableRepositoryRoles) {
+      fetchAvailablePermissionsIfNeeded(repositoryRolesLink, repositoryVerbsLink);
+    }
+  });
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-ui/ui-webapp/src/users/components/apiKeys/AddApiKey.tsx:67]: Great, thanks for this hint :-) Hope I will remember next time!

File: scm-ui/ui-webapp/src/users/components/apiKeys/ApiKeyCreatedModal.tsx [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+import { useTranslation } from "react-i18next";
+import styled from "styled-components";
+
+type Props = {
+  addedKey: string;
+  close: () => void;
+};
+
+const KeyArea = styled.textarea`
+  white-space: nowrap;
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-ui/ui-webapp/src/users/components/apiKeys/ApiKeyCreatedModal.tsx:36]: :+1:

File: scm-ui/ui-webapp/src/users/components/apiKeys/ApiKeyCreatedModal.tsx [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+      <p>
+        <b>{t("apiKey.modal.text2")}</b>
+      </p>
+      <hr />
+      <div className={"columns"}>
+        <div className={"column is-11"}>
+          <KeyArea wrap={"soft"} ref={keyRef} className={"input"} value={addedKey} />
+        </div>
+        <div className={"column is-1"}>
+          <Icon className={"is-hidden-mobile fa-2x"} name={copied ? "clipboard-check" : "clipboard"} title={t("apiKey.modal.clipboard")} onClick={copy} />
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-ui/ui-webapp/src/users/components/apiKeys/ApiKeyCreatedModal.tsx:64]: Yeah, got it :-)

File: scm-ui/ui-webapp/src/users/components/apiKeys/ApiKeyEntry.tsx [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  apiKey: ApiKey;
+  onDelete: (link: string) => void;
+};
+
+export const ApiKeyEntry: FC<Props> = ({ apiKey, onDelete }) => {
+  const [t] = useTranslation("users");
+  let deleteButton;
+  if (apiKey?._links?.delete) {
+    deleteButton = (
+      <a className="level-item" onClick={() => onDelete((apiKey._links.delete as Link).href)}>
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-ui/ui-webapp/src/users/components/apiKeys/ApiKeyEntry.tsx:41]: Maybe :-)

File: scm-webapp/src/main/java/sonia/scm/security/ApiKeyCollection.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+class ApiKeyCollection {
+  @XmlElement(name = "key")
+  private Collection<ApiKeyWithPassphrase> keys;
+
+  public ApiKeyCollection add(ApiKeyWithPassphrase key) {
+    Collection<ApiKeyWithPassphrase> newKeys;
+    if (keys == null) {
+      newKeys = new ArrayList<>();
+    } else {
+      newKeys = new ArrayList<>(keys.size() + 1);
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-webapp/src/main/java/sonia/scm/security/ApiKeyCollection.java:56]: Sure: Normally an ArrayList is created with an array of size 10. When you add more elements, this array has to be copied. Here we know exactly, how many keys we need. In this case this should be of no real relevance.
Nonetheless it is inconsistent that I create an empty normal ArrayList, when the keys are not set, yet. I'll fix this.

File: scm-webapp/src/main/java/sonia/scm/security/ScmWildcardPermission.java (hunk 2 above)
Comment (pfeuffer) [reply scm-webapp/src/main/java/sonia/scm/security/ScmWildcardPermission.java:40]: Tried it :-)

File: scm-webapp/src/test/java/sonia/scm/security/ApiKeyTokenHandlerTest.java (hunk 3 above)
Comment (pfeuffer) [reply scm-webapp/src/test/java/sonia/scm/security/ApiKeyTokenHandlerTest.java:43]: Definitely not, sorry

File: scm-webapp/src/test/java/sonia/scm/security/ScopesTest.java (hunk 4 above)
Comment (pfeuffer) [reply scm-webapp/src/test/java/sonia/scm/security/ScopesTest.java:74]: Ups, one of my favorite ones.

File: scm-webapp/src/main/java/sonia/scm/api/v2/resources/ApiKeyResource.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+      mediaType = VndMediaType.ERROR_TYPE,
+      schema = @Schema(implementation = ErrorDto.class)
+    ))
+  public HalRepresentation getForCurrentUser() {
+    return apiKeyCollectionMapper.map(apiKeyService.getKeys());
+  }
+
+  @GET
+  @Path("{id}")
+  @Produces(VndMediaType.API_KEY_COLLECTION)
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-webapp/src/main/java/sonia/scm/api/v2/resources/ApiKeyResource.java:96]: Definitely

File: scm-ui/ui-webapp/src/users/components/apiKeys/AddApiKey.tsx [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+  };
+};
+
+const mapDispatchToProps = (dispatch: any) => {
+  return {
+    fetchAvailablePermissionsIfNeeded: (repositoryRolesLink: string, repositoryVerbsLink: string) => {
+      dispatch(fetchAvailablePermissionsIfNeeded(repositoryRolesLink, repositoryVerbsLink));
+    }
+  };
+};
```
[the remark this answers is not in the document]
Comment (pfeuffer) [reply scm-ui/ui-webapp/src/users/components/apiKeys/AddApiKey.tsx:143]: Cannot resolve the imports. Do we have an outdated version of redux? Would leave it this way for now.
Thx for the hint, nonethelss!

**Pull request merged**